dnfdaemon-server/main.cpp

```diff
@@ -22,17 +22,40 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 #include "session_manager.hpp"
 
 #include <sdbus-c++/sdbus-c++.h>
+#include <signal.h>
 
 #include <iostream>
 
-static std::unique_ptr<SessionManager> session_manager = nullptr;
+static std::mutex session_manager_ptr_mutex;
+static std::atomic<SessionManager *> session_manager_ptr{nullptr};
+
+void sigint_handler([[maybe_unused]] int signum) {
+    // make sure that session_manager.shut_down() is called only once
+    SessionManager * local_session_manager_ptr{nullptr};
+    {
+        std::lock_guard<std::mutex> lock(session_manager_ptr_mutex);
+        local_session_manager_ptr = session_manager_ptr.load();
+        session_manager_ptr = nullptr;
+    }
+    if (local_session_manager_ptr != nullptr) {
+        local_session_manager_ptr->shut_down();
+    }
```

Comment:
What is the reason for the temporary `local_session_manager_ptr`? The mutex should already guarantee you're calling shutdown only once. Why not just:
```C++
std::lock_guard<std::mutex> lock(session_manager_ptr_mutex);
if (session_manager_ptr) {
    session_manager_ptr->shut_down();
    session_manager_ptr = nullptr;
}
```